Library scan: log summary and show popup

ronso0 opened the pull request:
This emphasizes the logged scan summary and allows to show a respective popup, both for automatic scan at startup and manual scan.

Fixes #10720 

For me, this is a step towards being able to manually find successors for moved missing tracks, may also be helpful for the _New Tracks_ filter/feature #13526.

```
info [LibraryScanner 1] ----------------------------------------------
info [LibraryScanner 1] Library scan finished after 963 ms
info [LibraryScanner 1]  1235 unchanged directories
info [LibraryScanner 1]  197 changed/added directories
info [LibraryScanner 1]  0 tracks verified from changed/added directories
info [LibraryScanner 1]  0 new tracks
info [LibraryScanner 1]  0 moved tracks
info [LibraryScanner 1]  0 missing tracks
info [LibraryScanner 1] ----------------------------------------------
```
![image](https://github.com/mixxxdj/mixxx/assets/5934199/d9ed13eb-976b-4d87-af3a-6a5bbd453d8a)
if nothing changed it displays only duration and ""No changes were detected."


TODO
~~Find an alternative for `int LibraryScanner::cleanUpScan()` to get the number of relocated tracks.~~ :heavy_check_mark:

ronso0 committed:
Library: improve logged scan summary, also log new and missing tracks

ronso0 committed:
Library: show scan summary popup

ronso0 commented:
> Find an alternative for int LibraryScanner::cleanUpScan() to get the number of relocated tracks.

Simply storing that in an LibraryScanner int member seems fine to me (reset when scan is started, no need to bother scannerGlobal with this).
Wdyt?

JoergAtGithub commented:
> show a respective popup, both for automatic scan at startup and manual scan

It's a nice addition to get a feedback, when I triggered the scan amnually. But I would no like to get this pop-up, when I start Mixxx.

ronso0 commented:
Funny thing is, I moved the connection to CoreServices so we can _always_ get the popup.

Why do you consider the info popup disturbing / not helpful for the autorun scan during  startup?

ronso0 commented on src/library/dao/trackdao.cpp:
```diff
@@ -1768,27 +1788,27 @@ void TrackDAO::markUnverifiedTracksAsDeleted() {
-    } // namespace
+    return matchLength;
+}
+} // namespace
```
This is only fixing indentation, pre-commit

ronso0 commented:
friendly ping @JoergAtGithub 
> Why do you consider the info popup disturbing / not helpful for the autorun scan during startup?

We _could_ make the report optional when scanning during startup, but I'd rather avoid that sort of complification / extra pref option.

JoergAtGithub commented:
In my opinion a dialog should only appear, if:

- it's the response to a users action (like clicking on rescan library here)
- an error occurs

but on startup it is always anoying to click dialogs away. Normal information should be visualize unobtrusive in the UI, in a way that does not require user action to continue.

ronso0 committed:
Library scan: don't show summary when run during startup

ronso0 commented:
The last commit disables the report/summary dialog for the auto scan run during startup.

ronso0 commented:
I'm a bit puzzled why the "changed/added" directory count is not 0 repeatedly even if the dirs haven't been modified.
(no regression btw)

ronso0 commented:
> I'm a bit puzzled why the "changed/added" directory count is not 0 repeatedly even if the dirs haven't been modified. (no regression btw)

Turns out it's just the confusing naming scanned vs. changed/added.
"scanned directories" are the directories where no previous hash exists (??) or prev hash != new hash (even though there was absolutely nothing even touching the dirs in between the scans :shrug: )

I adjusted the dialog and the logged lines. All good now!

ronso0 commented:
> Is it the intended behavior, that I get negative values, if I: Delete a file -> Rescan -> Restore the file -> Rescan

Thanks for your thourough testing!
This is not expected, but the numerical result of the TrackDAO/Scanner methods we had at our disposal:
```
  previously existing track locations
+ new tracks (= added tracks - relocated tracks)
- existing track locations
```

In order to reliably report rediscovered tracks I'll add a method to TrackDAO to get missing track locations.

ronso0 committed:
TrackDAO: add getAllMissingTrackLocations()

ronso0 committed:
Library scan: report rediscovered tracks, fix report of missing tracks

ronso0 commented:
The report now looks like this
![image](https://github.com/user-attachments/assets/a414c09c-3270-40cb-abf0-e1d539243c4e)

We _could_ calculate **new** missing tracks but I don't think that's worth it.

ronso0 commented:
> It would make sense to print the scan results into the log output at startup, if Mixxx is started in developer mode.

The report is always logged, just the dialog is now suppressed for the scan during startup.

daschuer commented:
I have not tested this yet and I have scanning during start up disabled so I probably not fully qualified to comment. But here some thoughts:

I think we have a scanning dialog that pops up when the scan takes longer. There is a risk that mixxx is locked when using the library during scanning. So I would argue the user is disturbed anyway by the startup scan. 

On the other hand I can confirm that the scan during startup is annoying and a "useless" dialog would be more annoying. 

How about showing the dialog only if new tracks are found or some are missing in the startup case? 
For my use case it would be also nice to put new tracks in an "Incoming" crate. 

What do you think?

ronso0 commented:
> I think we have a scanning dialog that pops up when the scan takes longer. There is a risk that mixxx is locked when using the library during scanning. So I would argue the user is disturbed anyway by the startup scan.

Exactly, the scan progress dialog pops up if the scan takes longer than 2s.
(fwiw during testing with auto-scan on startup I see the -empty- dialog artefact which never disappears, have to close it #12283)
Yes, the scan may be disturbing if one has new tracks on each start (many of them or a slow computer). And the scan during start with a fresh profile (new music dirs with lots of files) can lead to #11738 (which is why I have a delayed scan fix stashed).

ronso0 commented:
> How about showing the dialog only if new tracks are found or some are missing in the startup case?

Yes, that wold make sense. new tracks, new missing tracks, track rediscovered.

ronso0 commented:
> For my use case it would be also nice to put new tracks in an "Incoming" crate.

The "New tracks" feature / crate would make sense. How / when tracks are removed needs to be discussed. Automatically when a rating is set? And/or when they're sorted into playlists or crates?

I appreciate the idea, though it's unrelated to this UX fix/feature. I remember we briefly discussed this recently, is there already an Issue for tracking it?

ronso0 commented:
I think the report can be optimized. IMO this is sufficient:
> Scan took 1181 ms.
> 
>1 new tracks found
0 moved tracks detected
1 tracks are missing
0 tracks have been rediscovered
>
> 16066 tracks in total

**ToDo**
* clarify new tracks (subtract moved tracks)
* count new missing tracks (change discovered during this scan)

daschuer commented:
> The "New tracks" feature ...

My current use case would work perfectly if the tracks stick in the "Incoming" crate, just like in my email in box.

Normally I sort them from there and than weeks later I delete them they are no longer considered as "New".

daschuer commented:
We have this one: 
https://github.com/mixxxdj/mixxx/issues/13403

I will file new one for the scanner part.

daschuer commented:
https://github.com/mixxxdj/mixxx/issues/13526

ronso0 commented:
The report dialog now only displays relevant information:
new / moved / new missing / rediscovered tracks only if there are any.

I'll work on a fixup to show the report dialog during startup only if changes have been detected.

ronso0 committed:
Library scan: don't show report dialog when auto-scan detected no changes

ronso0 commented:
> I'll work on a fixup to show the report dialog during startup only if changes have been detected.

done.

ronso0 committed:
fixup! Library: improve logged scan summary, also log new and missing tracks

daschuer commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -422,27 +428,63 @@ void LibraryScanner::slotFinishUnhashedScan() {
+            numRediscoveredTracks++;
+        }
+    }
+    const auto missingTracks = m_trackDao.getAllMissingTrackLocations();
```
I have no idea how this performs and whether optimization is beneficial:
An idea for optimizion would be to move the compare loop into getAllMissingTrackLocations and get around the temporary QSet(). This will require a lot more compares though.

ronso0 replied:
> whether optimization is beneficial

Well, depends on if there are many tracks missing ('many' like 1.000 or 10.000). In case the library is that huge, there will be a (blocking) progress dialog anyway and IMHO +- a few milliseconds don't really make a difference then.

I prefer to have a clean QSet to work with here.
We could also use QSet::intersect(QSet&) and QSet::subtract(QSet&) to get rediscovered and new missing tracks
This would look cleaner but I wonder if the extra QSets are worth it.

ronso0 replied:
A quick benchmark with PerfrmanceTimer:
`TrackDAO - Collecting 16026 track locations took 23 ms`

daschuer reviewed (commented):
Some comments. Not yet tested.

ronso0 committed:
TrackDAO: add helper collectTrackLocations(), use FwdSqlQuery

ronso0 commented on src/library/dao/trackdao.cpp:
```diff
@@ -263,8 +263,51 @@ QList<TrackId> TrackDAO::resolveTrackIds(
+    while (query.next()) {
+        locations.insert(query.value(locationColumn).toString());
+    }
+    return locations;
```
thanks, done.

ronso0 committed:
TrackDAO/PlaylistDAO: use table columns constants

ronso0 committed:
fixup! Library scan: report rediscovered tracks, fix report of missing tracks

ronso0 committed:
Merge remote-tracking branch 'mixxx/main' into lib-scan-summary-etc

ronso0 committed:
fixup! Library scan: don't show report dialog when auto-scan detected no changes

ronso0 committed:
fixup! Library: improve logged scan summary, also log new and missing tracks

ronso0 committed:
Merge remote-tracking branch 'mixxx/main' into lib-scan-summary-etc

acolombier commented on src/library/dao/trackdao.cpp:
```diff
@@ -1905,6 +1940,7 @@ bool TrackDAO::detectMovedTracks(
             const auto nextSuffixMatch =
                     matchStringSuffix(nextTrackLocation, oldTrackLocation);
             DEBUG_ASSERT(nextSuffixMatch >= filename.length());
+            // document this
```
Any chance you can address this? Or maybe remove this comment otherwise

ronso0 replied:
I added the documentation.

acolombier commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -575,16 +618,20 @@ void LibraryScanner::slotAddNewTrack(const QString& trackPath) {
-        // The track's actual location might differ from the
-        // given trackPath
+        // The track's actual location might differ from the given trackPath.
+        // TODO why?
```
Did you find out? If not do you want to remove this `TODO`?

acolombier commented on src/library/scanner/libraryscanner.cpp:
```diff
@@ -575,16 +618,20 @@ void LibraryScanner::slotAddNewTrack(const QString& trackPath) {
         const QString trackLocation(pTrack->getLocation());
         // Acknowledge successful track addition
         if (m_scannerGlobal) {
+            // TODO emit pTrack, or use a new signal to store the track id
```
Is this still left `TODO`?

acolombier commented on src/library/scanner/scannerglobal.h:
```diff
@@ -144,6 +145,9 @@ class ScannerGlobal {
     }
     void trackAdded(const QString& trackLocation) {
         m_addedTracks << trackLocation;
+        // TODO Collect TrackIds?
```
Is this still left `TODO`?

acolombier commented on src/library/trackcollectionmanager.cpp:
```diff
@@ -84,6 +85,11 @@ TrackCollectionManager::TrackCollectionManager(
+                &LibraryScanner::scanSummary,
+                this,
+                &TrackCollectionManager::libraryScanSummary,
+                /*signal-to-signal*/ Qt::DirectConnection);
```
What's the meaning of `signal-to-signal`?

ronso0 replied:
I think this pattern has been used to indicate we don't connect to a slot but to a signal (signal forward essentially)

acolombier reviewed (commented):
Code LGTM, just nitpicking on comment cruft. Haven't had a chance to test yet. Could you address the `pre-commit` as well please?

ronso0 committed:
fixup! TrackDAO/PlaylistDAO: use table columns constants

ronso0 committed:
remove TODOs/ideas

ronso0 commented:
Thanks for taking a look!

Pre-commit should be happy now.
I removed the TODOs you mentioned.

ronso0 committed:
TrackDAO: document ranking of potential missing track successors

ronso0 committed:
Merge remote-tracking branch 'mixxx/main' into lib-scan-summary-etc

ronso0 committed:
Library scan: use comfortable duration format

daschuer reviewed (approved):
LGTM and works good. Thank you.

fwcd commented:
Just updated to the latest snapshot and stumbled upon this dialog. At the risk of feeling a little https://xkcd.com/1172, is there a setting for disabling this? I occasionally add music and rescan during a live set (via the keyboard shortcut) and having a modal popup show up is a bit obtrusive IMO.

PS. If Mixxx had some form of notification system, I'd imagine this being a great use case for that. Or maybe there's either a Qt or some other cross-platform way of sending desktop notifications?

ronso0 commented:
Nope, there' no setting.
Until we have some kind of notification service (we have a tracking issue for that but i just can't find it)
we could add a setting (a drop-down)
**Show Scan Summary**
* always
* not when playing
* never
